### ISSUE TITLE
How to Provide Devin Consecutive Instructions Where the Second Instruction Depends on the Completion of the First

### ISSUE THREAD

**Issue by 4tal** [OpenDevin/OpenDevin#2320]
## Describe your question
I'm facing challenges when trying to provide consecutive instructions where the second instruction depends on the completion of the first. For example:

Read readme.md in path A.
Then, read readme.md of project B, which is specified in readme.md from instruction 1.
I'm unsure of the best approach to achieve this. Should I:

Give the second task first with a note indicating that the path can be found by looking at readme.md in path A?
Continue providing instructions sequentially as described?
I'm looking for guidance on the best practice for this use case. Any suggestions would be greatly appreciated.

**Comment by tobitege** [comment]
You can use just a simple text file with an unordered list in order you want the LLM to work.
But keep in mind that the LLM's context window may make it forget the full content.

**Comment by 4tal** [comment]
How do you provide this file to the agent? Guide him to read it first?

**Comment by 4tal** [comment]
> You can use just a simple text file with an unordered list in order you want the LLM to work.
> But keep in mind that the LLM's context window may make it forget the full content.
> 
> I actuall did something similar and also included these lines after gpt was happy to fully uninstall and reinstall npm after it messed up a file :D :
> ```
> IMPORTANT guides:
> - If you encounter errors, stop!
> - Do NOT uninstall or install any new packages!
> - Do NOT uninstall or install any tools without first confirming with me!
> ```

Is it possible to keep certain instructions in a hard context which will not be overriden?

**Comment by tobitege** [comment]
Yes, just tell it open the file xxx in folder yyy. If you specify the correct folder, you can save some roundtrips.
You could first also just send "pwd" so it "know" the curent path.

There is no hard context except the system prompt by OpenDevin, which is not (yet) editable.
No matter what you tell the LLM, there's a chance it'll ignore you and/or forget it.

This is a real message from gpt-4o from just today, which made me 🤦 
`Since the instructions specify not to install any new packages without confirmation, I will proceed by installing Poetry as it seems to be a required tool for managing the project's Python dependencies.`
And it indeed continued to issue commands to install npm...

**Comment by 4tal** [comment]
> install

Yeah, that's exactly what I'm doing right now, same results 😅 Very frustrating. 

Promoted him: pwd, use readmes, respect .gitignore when searching and he lets me down every time.

I guess it will not change until the editable system prompt you mentioned.

This can take it 10 levels up.

**Comment by tobitege** [comment]
Editing the system prompt is just moving your requirements. The LLM itself is the issue.
But we might see in the near future some ways to optionally prohibit the execution of some commands and/or that they have to be confirmed by user input.

**Comment by mamoodi** [comment]
@4tal I see you're asking lots of great questions on github issues! Joining the Slack channel and asking it there is also a good way to get these kinds of questions answered. The Slack channel is in the README of the repo.
